### ISSUE TITLE
.not() silently fails with complex selectors

### ISSUE THREAD

**Issue by aweary** [airbnb/enzyme#1078]
I'm doing some work on getting an actual CSS parser and traversal utility for RST, and noticed this issue. If you try to use a complex selector with `.not()` like so:

```js
wrapper.find('.foo').not('.bar + div')
```

It will silently fail. This is because we use `buildPredicate` and it does not check for complex selectors. We should either support complex selectors with `buildPredicate` or explicitly fail.

**Comment by aweary** [comment]
This is also true of any other method that uses `buildPredicate`, such as `is()` or `filter()`. @ljharb do you think it makes sense to support complex selectors with those methods? I think limiting them to simple compound selectors is reasonable.

**Comment by ljharb** [comment]
Ideally we'd support complex ones; but I agree that if we don't it should fail loudly.

**Comment by aweary** [comment]
I think I can find a way to support complex selectors. Otherwise, I'll add a check + error.

**Comment by aweary** [comment]
I think the biggest problem with supporting complex selectors with filtering methods like `not` or `filter` is that it will require us to traverse the whole tree for each node. For example, if you have:

```jsx
<div>
 <div className="bar">
    <span>Foo</span>
  </div>
  <span>Bar</span>
</div>
```

With the query:

```js
wrapper.find('span').filter('.bar > span')
```

In order to know whether the `span` is a child of `.bar` we need to move up the RST, which isn't possible since we don't have parent pointers. That means we need to start top-down for each node to find its parent. Limiting complex selectors to `find` means we can avoid having extra logic for this kind of traversal.

**Comment by ljharb** [comment]
Closed in #1086
